services/src/main/java/org/keycloak/protocol/ciba/decoupledauthn/DecoupledAuthenticationProviderBase.java

```diff
@@ -0,0 +1,263 @@
+
+        Response response = verifyDecoupledAuthnResult();
+        if (response != null) return response;
+
+        setupSessions();
+
+        persistDecoupledAuthenticationResult(DecoupledAuthStatus.SUCCEEDED);
+
+        return cors.builder(Response.ok("", MediaType.APPLICATION_JSON_TYPE)).build();
+    }
+
+    private void setupSessions() {
+        RootAuthenticationSessionModel rootAuthSession = session.authenticationSessions().createRootAuthenticationSession(realm, getUserSessionIdWillBeCreated());
+        // here Client Model of CD(Consumption Device) needs to be used to bind its Client Session with User Session.
+        AuthenticationSessionModel authSession = rootAuthSession.createAuthenticationSession(client);
+
+        authSession.setProtocol(OIDCLoginProtocol.LOGIN_PROTOCOL);
+        authSession.setAction(AuthenticatedClientSessionModel.Action.AUTHENTICATE.name());
+        authSession.setClientNote(OIDCLoginProtocol.ISSUER, Urls.realmIssuer(session.getContext().getUri().getBaseUri(), realm.getName()));
+        authSession.setClientNote(OIDCLoginProtocol.SCOPE_PARAM, getScope());
+        logger.info(" specified scopes in backchannel authentication endpoint = " + getScope());
+
+        // authentication
+        AuthenticationFlowModel flow = AuthenticationFlowResolver.resolveCIBAFlow(authSession);
+        String flowId = flow.getId();
+        AuthenticationProcessor processor = new AuthenticationProcessor();
+        processor.setAuthenticationSession(authSession)
+                .setFlowId(flowId)
+                .setConnection(clientConnection)
+                .setEventBuilder(event)
+                .setRealm(realm)
+                .setSession(session)
+                .setUriInfo(session.getContext().getUri())
+                .setRequest(httpRequest);
+
+        processor.authenticateOnly();
+        processor.evaluateRequiredActionTriggers();
+        UserModel user = authSession.getAuthenticatedUser();
+        if (user.getRequiredActionsStream().count() > 0) {
+            event.error(Errors.RESOLVE_REQUIRED_ACTIONS);
+            throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_GRANT, "Account is not fully set up", Response.Status.BAD_REQUEST);
+        }
+
+        AuthenticationManager.setClientScopesInSession(authSession);
+
+        processor.attachSession();
+        UserSessionModel userSession = processor.getUserSession();
+        if (userSession == null) {
+            event.error(Errors.USER_SESSION_NOT_FOUND);
+            throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_GRANT, "User session is not found", Response.Status.BAD_REQUEST);
+        } else {
+            logger.infof(" user session id =  %s, username = %s", userSession.getId(), userSession.getUser().getUsername());
+        }
+        updateUserSessionFromClientAuth(userSession);
+
+        logger.infof("  Created User Session's id                            = %s", userSession.getId());
```

Comment:
Too verbose? 

Could you please check other logging messages in other places that fit into the same problem?

Reply:
@pedroigor I will review logging messages and remove them.

Reply:
@pedroigor I think these verbose logging message might be beneficial for debugging (check whether CIBA works well). So, how about outputting these messages as trace log facility like `ClientPolicyLogger` of Client Policies?
If so, `ClientPolicyLogger` itself might be used for CIBA.

Reply:
I think log using trace lavel makes sense.